File: app/src/main/java/com/ivanovsky/passnotes/domain/interactor/newdb/NewDatabaseInteractor.java
package com.ivanovsky.passnotes.domain.interactor.newdb;

import com.ivanovsky.passnotes.data.ObserverBus;
import com.ivanovsky.passnotes.data.entity.FileDescriptor;
import com.ivanovsky.passnotes.data.entity.OperationResult;
import com.ivanovsky.passnotes.data.entity.OperationError;
import com.ivanovsky.passnotes.data.entity.UsedFile;
import com.ivanovsky.passnotes.data.repository.EncryptedDatabaseRepository;
import com.ivanovsky.passnotes.data.repository.UsedFileRepository;
import com.ivanovsky.passnotes.data.repository.encdb.EncryptedDatabase;
import com.ivanovsky.passnotes.data.repository.encdb.exception.EncryptedDatabaseException;
import com.ivanovsky.passnotes.data.repository.file.FileSystemProvider;
import com.ivanovsky.passnotes.data.repository.file.FileSystemResolver;
import com.ivanovsky.passnotes.data.repository.keepass.KeepassDatabaseKey;
import com.ivanovsky.passnotes.injection.Injector;
import com.ivanovsky.passnotes.util.Logger;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

import static com.ivanovsky.passnotes.data.entity.OperationError.newFileIsAlreadyExistsError;
import static com.ivanovsky.passnotes.data.entity.OperationError.newGenericError;

public class NewDatabaseInteractor {

	private final EncryptedDatabaseRepository dbRepository;
	private final UsedFileRepository usedFileRepository;
	private final ObserverBus observerBus;
	private final FileSystemResolver fileSystemResolver;

	public NewDatabaseInteractor(EncryptedDatabaseRepository dbRepository,
								 UsedFileRepository usedFileRepository,
								 FileSystemResolver fileSystemResolver,
								 ObserverBus observerBus) {
		this.dbRepository = dbRepository;
		this.usedFileRepository = usedFileRepository;
		this.observerBus = observerBus;
		this.fileSystemResolver = fileSystemResolver;
	}

	public Single<OperationResult<Boolean>> createNewDatabaseAndOpen(KeepassDatabaseKey key,
																	 FileDescriptor file) {
		return Single.fromCallable(() -> createNewDatabaseAndOpenAsync(key, file))
				.subscribeOn(Schedulers.newThread())
				.observeOn(AndroidSchedulers.mainThread());
	}

	private OperationResult<Boolean> createNewDatabaseAndOpenAsync(KeepassDatabaseKey key,
																   FileDescriptor file) {
		OperationResult<Boolean> result = new OperationResult<>();

		FileSystemProvider provider = fileSystemResolver.resolveProvider(file.getFsType());
		if (!provider.exists(file)) {
			if (dbRepository.createNew(key, file)) {
				UsedFile usedFile = new UsedFile();

				usedFile.setFilePath(file.getPath());
				usedFile.setFileUid(file.getUid());
				usedFile.setFsType(file.getFsType());
				usedFile.setLastAccessTime(System.currentTimeMillis());

				usedFileRepository.insert(usedFile);

				observerBus.notifyUsedFileDataSetChanged();

				try {
					EncryptedDatabase db = dbRepository.open(key, file);

					Injector.getInstance().createEncryptedDatabaseComponent(db);

					result.setResult(true);
				} catch (EncryptedDatabaseException e) {
					Logger.printStackTrace(e);

					result.setError(newGenericError(OperationError.MESSAGE_UNKNOWN_ERROR, e));
				}
			} else {
				result.setError(newGenericError(OperationError.MESSAGE_UNKNOWN_ERROR));
			}
		} else {
			result.setError(newFileIsAlreadyExistsError());
		}

		return result;
	}
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/domain/interactor/newdb/NewDatabaseInteractor.kt
package com.ivanovsky.passnotes.domain.interactor.newdb

import com.ivanovsky.passnotes.data.entity.FileDescriptor
import com.ivanovsky.passnotes.data.entity.OperationError.*
import com.ivanovsky.passnotes.data.entity.OperationResult
import com.ivanovsky.passnotes.data.repository.EncryptedDatabaseRepository
import com.ivanovsky.passnotes.data.repository.UsedFileRepository
import com.ivanovsky.passnotes.data.repository.file.FSOptions
import com.ivanovsky.passnotes.data.repository.file.FileSystemResolver
import com.ivanovsky.passnotes.data.repository.keepass.KeepassDatabaseKey
import com.ivanovsky.passnotes.domain.DispatcherProvider
import com.ivanovsky.passnotes.domain.usecases.AddTemplatesUseCase
import com.ivanovsky.passnotes.extensions.toUsedFile
import kotlinx.coroutines.withContext

class NewDatabaseInteractor(
    private val dbRepo: EncryptedDatabaseRepository,
    private val usedFileRepository: UsedFileRepository,
    private val fileSystemResolver: FileSystemResolver,
    private val addTemplatesUseCase: AddTemplatesUseCase,
    private val dispatchers: DispatcherProvider
) {

    suspend fun createNewDatabaseAndOpen(
        key: KeepassDatabaseKey,
        file: FileDescriptor,
        isAddTemplates: Boolean
    ): OperationResult<Boolean> {
        return withContext(dispatchers.IO) {
            val provider = fileSystemResolver.resolveProvider(file.fsAuthority)

            val existsResult = provider.exists(file)
            if (existsResult.isFailed) {
                return@withContext existsResult.takeError()
            }

            val isExists = existsResult.obj
            if (isExists) {
                return@withContext OperationResult.error(newFileIsAlreadyExistsError())
            }

            val creationResult = dbRepo.createNew(key, file, isAddTemplates)
            if (creationResult.isFailed) {
                return@withContext creationResult.takeError()
            } else if (creationResult.isDeferred) {
                return@withContext OperationResult.error(
                    newFileAccessError(MESSAGE_DEFERRED_OPERATIONS_ARE_NOT_SUPPORTED)
                )
            }

            val openResult = dbRepo.open(key, file, FSOptions.DEFAULT)
            if (openResult.isFailed) {
                return@withContext openResult.takeError()
            }

            val getFileResult = getFile(file)
            if (getFileResult.isFailed) {
                return@withContext getFileResult.takeError()
            }

            val newFile = getFileResult.obj
            val time = System.currentTimeMillis()
            val usedFile = newFile.toUsedFile(
                addedTime = time,
                lastAccessTime = time
            )

            usedFileRepository.insert(usedFile)

            OperationResult.success(true)
        }
    }

    private fun getFile(file: FileDescriptor): OperationResult<FileDescriptor> {
        val provider = fileSystemResolver.resolveProvider(file.fsAuthority)
        val getFileResult = provider.getFile(file.path, FSOptions.DEFAULT)

        return when {
            getFileResult.isSucceeded -> getFileResult
            getFileResult.isDeferred -> OperationResult.error(
                newFileAccessError(MESSAGE_DEFERRED_OPERATIONS_ARE_NOT_SUPPORTED)
            )
            else -> getFileResult.takeError()
        }
    }
}